### PR TITLE
feature/006-support-dotfiles

### PR DESCRIPTION
**Pull request by mikeerickson**
- [x] Add support for global `.phpunit-watcher.yml` files up the directory tree
- [x] Once a `.phpunit-watcher.yml` is found, its contents will be used

### CODE REVIEW

File: src/WatcherCommand.php [hunk 1]
```diff
@@ -8,9 +8,12 @@
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Output\OutputInterface;
+use function var_dump;
 
 class WatcherCommand extends Command
 {
+    const PHPUNIT_WATCH_CONFIG_FILENAME = "/.phpunit-watcher.yml";
```
Comment (freekmurze) [inline src/WatcherCommand.php]: The class is kinda small, this constant is only being used in one method. That's why I think that using a constant here is a bit overkill. Use a regular variable to hold this value in the `getOptionsConfigFile` method itself.

File: .idea/vcs.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (freekmurze) [inline .idea/vcs.xml]: Remove this file.

File: src/WatcherCommand.php [hunk 3]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
         return $options;
     }
 
+    protected function getOptionsConfigFile()
```
Comment (freekmurze) [inline src/WatcherCommand.php]: I feel that this method can be greatly simplified.

Rename this to `getOptionsFromFile`, add an `array` return typehint.

File: src/WatcherCommand.php [hunk 4]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
...
+        $configPath = getcwd();
+        $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+
+        while (! file_exists($configFile)):
+            $configPath = dirname($configPath);
+            $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+            if ($configPath === '/') {
+                $configFile = self::PHPUNIT_WATCH_CONFIG_FILENAME;
+                break;
+            }
```
Comment (freekmurze) [inline src/WatcherCommand.php]: Inside the while, if a config file is found, immediately read it and return the config values.

File: src/WatcherCommand.php [hunk 5]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
...
+        endwhile;
+
+        return $configFile;
```
Comment (freekmurze) [inline src/WatcherCommand.php]: If you're outside this loop, it means that no config file was found. Just return an empty array.

File: src/WatcherCommand.php [hunk 6]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
...
+    }
+
     protected function getOptionsFromConfigFile(): array
     {
-        $configFile = getcwd().'/.phpunit-watcher.yml';
+        $configFile = $this->getOptionsConfigFile();
```
Comment (freekmurze) [inline src/WatcherCommand.php]: The code below this line will parse the file. Of course that should be done in the `getOptionsFromFile` now.

File: src/WatcherCommand.php [hunk 7]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
...
 
+    protected function getOptionsConfigFile()
+    {
+        $configPath = getcwd();
+        $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+
+        while (! file_exists($configFile)):
+            $configPath = dirname($configPath);
+            $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+            if ($configPath === '/') {
```
Comment (freekmurze) [inline src/WatcherCommand.php]: Checking on a `/`  string feels kinda brittle.

I'm thinking checking on the return value `pathinfo($configPath, PATHINFO_DIRNAME);` better conveys what you're trying to do.

File: src/WatcherCommand.php [hunk 8]
```diff
@@ -55,7 +75,7 @@ protected function displayOptions(array $options, InputInterface $input, OutputI
 
         $output->title('PHPUnit Watcher');
 
-        $output->text("Tests will be rerun when {$options['watch']['fileMask']} files are modified in");
+        $output->text("Tests will be rerun when {$options['watch']['fileMask']} files are modified in the following directories:\n");
```
Comment (freekmurze) [inline src/WatcherCommand.php]: Thanks for this, but I liked the short version better. Could you restore that?

**Comment by mikeerickson** [comment]
@freekmurze Everybody coding style is different, this is what makes contribution difficult.  I will review your comments and update the PR accordingly.

File: src/WatcherCommand.php [hunk 9]
```diff
@@ -38,9 +41,26 @@ protected function determineOptions(InputInterface $input): array
...
+        $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+
+        while (! file_exists($configFile)):
+            $configPath = dirname($configPath);
+            $configFile = $configPath . self::PHPUNIT_WATCH_CONFIG_FILENAME;
+            if ($configPath === '/') {
+                $configFile = self::PHPUNIT_WATCH_CONFIG_FILENAME;
+                break;
+            }
+        endwhile;
```
Comment (mdrost) [inline src/WatcherCommand.php]: IMHO `endwhile` should be only used inside code mixed with HTML.

File: src/WatcherCommand.php [hunk 10]
```diff
@@ -8,9 +8,12 @@
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Output\OutputInterface;
+use function var_dump;
```
Comment (mdrost) [inline src/WatcherCommand.php]: That `var_dump` looks like leftovers.

Comment (freekmurze) [inline src/WatcherCommand.php]: Yea, this line should be removed.

**Comment by freekmurze** [comment]
Thanks for the initial fix, I modified it a bit before merging it in.

**Pull request merged**